runtime/jailer.go

```diff
@@ -0,0 +1,403 @@
+		OCIBundlePath:  ociBundlePath,
+		UID:            jailerUID,
+		GID:            jailerGID,
+		RuncBinaryPath: config.JailerConfig.RuncBinaryPath,
+	}
+
+	return newRuncJailer(logger, cfg), nil
+}
+
+// jailer will allow modification and provide options to the the Firecracker VM
+// to allow for jailing. In addition, this will allow for given files to be exposed
+// to the jailed filesystem.
+type jailer interface {
+	// BuildJailedMachine will modify the firecracker.Config and provide
+	// firecracker.Opt to be passed into firecracker.NewMachine which will allow
+	// for the VM to be jailed.
+	BuildJailedMachine(*Config, *firecracker.Config, string) ([]firecracker.Opt, error)
+	// ExposeDeviceToJail will expose the given device provided by the snapshotter
+	// to the jailed filesystem
+	ExposeDeviceToJail(string) error
+}
+
+// noopJailer is a jailer that returns only successful responses and performs
+// no operations during calls
+type noopJailer struct{}
+
+func (j noopJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	return []firecracker.Opt{}, nil
+}
+
+func (j noopJailer) ExposeDeviceToJail(path string) error {
+	return nil
+}
+
+type runcJailConfig struct {
+	OCIBundlePath  string
+	UID            uint32
+	GID            uint32
+	RuncBinaryPath string
+}
+
+// runcJailer uses runc to set up a jailed environment for the Firecracker VM.
+type runcJailer struct {
+	logger *logrus.Entry
+
+	Config runcJailConfig
+}
+
+func newRuncJailer(logger *logrus.Entry, config runcJailConfig) *runcJailer {
+	return &runcJailer{
+		logger: logger,
+		Config: config,
+	}
+}
+
+// JailPath returns the base directory from where the jail binary will be ran
+// from
+func (j runcJailer) OCIBundlePath() string {
+	return j.Config.OCIBundlePath
+}
+
+// RootPath returns the root fs of the jailed system.
+func (j runcJailer) RootPath() string {
+	return filepath.Join(j.OCIBundlePath(), rootfsFolder)
+}
+
+func (j runcJailer) ContentsPath() string {
+	return filepath.Join(j.RootPath(), "var", "lib", "firecracker-containerd")
+}
+
+func (j *runcJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	handler := j.BuildHandler(cfg, &machineConfig.SocketPath, vmID)
+	fifoHandler := j.BuildLinkFifoHandler()
+	client := firecracker.NewClient(machineConfig.SocketPath, j.logger, machineConfig.Debug)
+
+	opts := []firecracker.Opt{
+		firecracker.WithProcessRunner(j.jailerCommand(vmID)),
+		firecracker.WithClient(client),
+		func(m *firecracker.Machine) {
+			m.Handlers.FcInit = m.Handlers.FcInit.Prepend(handler)
+			m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(firecracker.CreateLogFilesHandlerName, fifoHandler)
+		},
+	}
+
+	return opts, nil
+}
+
+// BuildHandler will link the necessary files except for the fifos due to the
+// fifos needing to be created. Also, this will create the proper device nodes
+// needed by Firecracker
```

Comment:
I think this comment could be a little more clear. Maybe "BuildHandler will populate the jail with the necessary files, which may be device nodes, hard links, or bind-mount targets." We can say something about the FIFOs being created elsewhere, or not. If so, I think we could use a clearer explanation than what's here currently.